Admin Panel to replace admin features in Hall of Heroes - 2

Alys opened the pull request:
Partial fix for HabitRPG/habitica-private#119

Continuation of https://github.com/HabitRPG/habitica/pull/12035

This contains all the commits from the previous PR, plus three new commits, mostly for fixes requested by paglias. 
The comments that this PR fixes start at https://github.com/HabitRPG/habitica/pull/12035#pullrequestreview-500422316

Ready for review (assuming tests pass).

Alys committed:
create Admin Panel page with initial content from Hall's admin section

Alys committed:
reorganise Admin Panel form and add more accordians

Alys committed:
add lastCron to fields returned by api.getHeroes

Alys committed:
improve timestamps and authentication section

Alys committed:
add party and quest info to Admin Panel, add party to heroAdminFields

Alys committed:
move Admin Panel menu item to top of menu, make invisible to non-admins

Alys committed:
remove code used for displaying all Heroes

Alys committed:
add avatar appearance and drops section in Admin Panel

Alys committed:
allow logged-in user to be the default hero loaded

Alys committed:
add time zones to timestamp/authentication section

Alys committed:
rename Items to Update Items

This will allow a new Items section to be added.

Alys committed:
add read-only Items display with button to copy data to Update Items section

Alys committed:
remove never-used allItemsPaths code that had been copied from Hall

Alys committed:
update tests for the attributes added to heroAdminFields

Alys committed:
supply names for items and also set information for gear/equipment

Alys committed:
remove code that loads subsections of content

We use enough of the content that it's easier to load it all and
access it through the content object, especially when we're looping
through different item types.

Alys committed:
add gear names and set details to Avatar Costume/Battle Gear section

Alys committed:
make the wiki URLs clickable and make minor item format improvements

Alys committed:
add gear sets for Check-In Incentives and animal ears and tails

Alys committed:
add gear set for Gold-Purchasable Quest Lines

Also merges the existing Mystery of the Masterclassers quest set into it.

Alys committed:
fix error with Kickstarter gear set and include wiki link

Alys committed:
improve description of check-in incentive gear set

Alys committed:
fix description of Items section

Alys committed:
fix lint warnings

Alys committed:
update another test for the attributes added to heroAdminFields

Alys committed:
allow "@" to be included when specifying Username to load

Alys committed:
create GetHeroParty API v3 route to fetch a given user's party data

Only some data from the party will be loaded (e.g., not private
data such as name, description).

Includes tests for the route.

See the next commit for front-end changes that use this.

Alys committed:
display data from a given user's party in admin panel

Only some data from the party will be loaded (e.g., not private
data such as name, description).

Also adds support for finding and displaying errors from the
user's data.

Alys committed:
Merge remote-tracking branch 'upstream/develop' into admin-panel-conflict-fixes

Also adjusts
test/api/v3/integration/hall/GET-hall_heroes_heroId.test.js
to use the heroFields array concept from
test/api/v3/integration/hall/PUT-hall_heores_heroId.test.js

Alys committed:
use new error handling method for other sections

- Time zone differences
- Cron bugs
- Privilege removal (mute/block) - not a bug but needs to be highlighted

Alys committed:
redirect non-admin users away from admin-only page (WIP)

This needs more work. Currently, admin users are also redirected
if they access the page by direct URL or after reload.

Alys committed:
clarify source of items from Check-In Incentives and Lunar Battle quests

Alys committed:
replace non-standard form fields with HTML forms

Alys committed:
Merge remote-tracking branch 'upstream/develop' into admin-panel-new

Alys committed:
add user's language, remove unused export blocks

Alys committed:
convert functions to filters: formatDate, formatTimeZone

Alys committed:
improve display of minutes portion of time zone in Admin Panel

Alys committed:
move basic details about user to a new component

Alys committed:
move Timestamp/Cron/Auth/etc details to a new component - WIP, has errors

The automatic expand and error warnings don't reset themselves when
you fetch data for a new user.

Alys committed:
replace non-standard form fields with HTML forms

Most of this was done in 26fdcbbee51937d39a149a20bf7ccd1cf674ca18

Alys committed:
move Timestamp/Cron/Auth/etc details to a new component (fixed)

Alys committed:
move Avatar and Drops section to a new component

Alys committed:
move Party and Quest section to a new component

Alys committed:
move Contributor Details to new component, add checkbox for admin, add preview

This adds a markdown-enabled preview of the Contributions textarea.

It also removes the code that automatically set contributor.admin
to true when the Tier was above 7.
That feature wasn't secure because the Tier can be accidentally
changed if you scroll while the cursor is over the Tier form field
(we accidentally demoted a Socialite once by doing that and if
we'd scrolled in the other direction we would have given her
admin privileges).

Instead there's now a checkbox for giving moderator-level privileges.
We'll want that anyway when we move to a system of selected
privileges for each admin instead of all admin privileges being
given to all mods/staff.

There's also a commented-out checkbox for giving Bailey CMS
privileges, for when we're ready to use that. The User model doesn't
yet have support for it.

Alys committed:
move Privileges and Gems section to a new component

Alys committed:
rename formatItems to getItemDescription; make other minor fixes

Alys committed:
add newsPoster Bailey CMS permission to User model and Admin Panel

Alys committed:
remove an outdated test description

This "pended" explanation probably wasn't needed after "x" was
removed from "describe" in 2ab76db27cbc2ea6c8e65a407a3fcc51fb015687

Alys committed:
move formatDate from mixins to filters

Alys committed:
make lint fixes

Alys committed:
remove development comments from hall.js

I'll be handling the TODO comment and I've left in my "XXX" marker
to remind me

Alys committed:
fix bug in Hall's castItemVal: mounts are null not false

Alys committed:
move Items section to a new component and delete Update Items section

The Update Items section is no longer needed because the new Items
component has in-place editing.

Alys committed:
remove unused imports

Alys committed:
add "secret" field to "Privileges, Gem Balance" section.

Also move the markdownPreview style from contributorDetails.vue to
index.vue since it's used in two components now.

Alys committed:
show non-Standard never-owned Pets and Mounts in Items section

Alys committed:
redirect non-admin users away from admin-only page

This completes the work started in commit a4f9c754ad3886466925de4d8ed9a5ac55d2f999

It now allows admins to access the page when coming from another
page on the site or from a direct link, including if the admin user
isn't logged in yet.

Alys committed:
display memberCount for party

Alys committed:
add secret.text field to Contributor Details

This is in addition to showing it in the Privileges section because
the secret text could be about either troublesome behaviour or
contributions.

Alys committed:
allow user to be loaded into Admin Panel via a URL

This includes:

- router config has a child route for the admin panel with a
Username/ID as a parameter
- loadHero code moved from top-level index page into a new
"user support" index page
- links in the Hall changed to point to admin panel route
- admin panel link added to admin section of user profile modal

Alys committed:
keep list of known titles on their own lines

Alys committed:
Merge remote-tracking branch 'upstream/develop' into admin-panel

Alys committed:
sort heroFields alphabetically

No actual changes.

Alys committed:
return all flags for use in Admin Panel and fix Hall tests for flags

Future Admin Panel changes will display more flags.

NB 'flags' wasn't in the tests before, even though two optional
flags were being fetched.
The tests weren't failing because the test users hadn't been given
data for those optional flags.

The primary reason for this change now is to fix the tests.

Alys committed:
show part of the API Token in the Admin Panel

Alys committed:
send full hero object into cronAndAuth.vue

This is a prelude to allowing this component to change the hero.

Alys committed:
split heroAdminFields string into two: one for fetching data and one for showing it

This is because apiToken must be fetched but not shown,
while apiTokenObscured is calculated (not fetched) and shown.

Alys committed:
let admin change a user's API Token

Alys committed:
restore sanity

Alys committed:
remove code to show obscured version of API Token

It will return with tighter permissions for viewing it.

Alys committed:
add Custom Day Start time (CDS) to Timestamps, Time Zone... section

Alys committed:
commit lint's automatic fixes - one for admin-panel changes in hall.js

The other fixes aren't related to this PR but I figured they may
as well go live.

Alys committed:
apply fixes from paglias's comments, excluding style/CSS changesd

The comments that this PR fixes start at
https://github.com/HabitRPG/habitica/pull/12035#pullrequestreview-500422316

Style fixes will be in a future commit.

Alys committed:
fix styles/CSS

Alys committed:
allow profile modal to close when using admin panel link

Also removes an empty components block.

Alys commented:
The comments that this PR fixes start at
https://github.com/HabitRPG/habitica/pull/12035#pullrequestreview-500422316

paglias commented:
Pushed to delta, tomorrow I'll make a final review!

paglias commented on website/client/src/components/admin-panel/user-support/index.vue:
```diff
@@ -0,0 +1,171 @@
+      this.$emit('changeUserIdentifier', id); // change user identifier in Admin Panel's form
+
+      const hero = await this.$store.dispatch('hall:getHero', { uuid: id });
+      this.hero = { ...hero };
```
...hero is not necessary when replacing the entire object

Alys replied:
that's fixed

paglias commented on website/client/src/router/index.js:
```diff
@@ -338,6 +342,22 @@ const router = new VueRouter({
+      path: '/admin-panel',
+      component: AdminPanelPage,
+      meta: { requiresAdmin: true },
+      children: [
```
if you Save and Clear data after loading an user in the admin panel the url doesn't change and clicking the Load User button again it refreshes the page

Alys replied:
So sorry, I only just noticed these review comments. I must have missed the email. :(

Yes, that behaviour is intentional.
"Save and Clear" for the contributor section is necessary because increasing the Tier will change other parts of the user (Gems, items, pets) and so you want to ensure that none of the old data remains.

However the User ID is kept in the search box and the URL doesn't change so that you can easily re-fetch the data for the same user, in case you want to check their new settings or make other changes to their account.

The reload is necessary because the vue router can't handle "moving" from one URL to the same URL. The same thing happens if you load a user and then click the "Load User" button a second time (e.g., if you want to back out of any changes you made). That's this code: https://github.com/HabitRPG/habitica/pull/12652/files#diff-7a3e28ebec99ac2c419cdc471d62b162R60-R75

I'm sure there's a nicer way to do all this but I messed around with it for hours and this was the best option that covered all cases. If it was for the general user population it would be worth spending more time on it, but I'm not sure it's worth it for only the staff and mods.

paglias replied:
> However the User ID is kept in the search box and the URL doesn't change so that you can easily re-fetch the data for the same user, in case you want to check their new settings or make other changes to their account.

I'd like this to be changed, I'm fine with keeping the user id in the search box but if we go back to the main page the url should be changed accordingly

> "Save and Clear" for the contributor section is necessary because increasing the Tier will change other parts of the user (Gems, items, pets) and so you want to ensure that none of the old data remains.

I'm fine if we want to keep the Save and Clear but I'm pretty sure the api route used to update the data returns everything for that user so you could take the response and Object.assign(this.hero, response) to update it (but if it's too much hassle, Save and clear is totally fine for me)

Alys committed:
prevent Admin Panel being used without new userSupport privilege

Also adds initial support for other contributor.priv privileges
and changes Debug Menu to add userSupport privilege

Alys committed:
don't do this: this.hero = { ...hero };

Alys committed:
enhance quest error messages

paglias commented on website/server/controllers/api-v3/hall.js:
```diff
@@ -310,16 +322,63 @@ api.updateHero = {
+api.getHeroParty = { // @TODO XXX add tests
+  method: 'GET',
+  url: '/hall/heroes/party/:groupId',
+  middlewares: [authWithHeaders(), ensureAdmin],
```
so for the routes here, we're adding the new privileges in the user schema but not using them? I mean all the routes here are checking contributor.admin, not one of the new privileges so they're still accessible from the api

Alys committed:
redirect to admin-panel home page when using "Save and Clear Data"

The user's ID / name is still in the form for easy refetching.

Alys committed:
create ensurePriv function, use in api.getHeroParty

Alys committed:
Merge remote-tracking branch 'upstream/develop' into admin-panel-2

paglias commented:
Looks like some tests are failing

Alys committed:
fix lint problems and integration tests

paglias commented:
Now that https://github.com/HabitRPG/habitica/pull/12627 has been merged we should call the `common:setTitle` action when the admin panel is loaded to make sure it's a proper title

paglias commented:
@Alys what's the status of this?

Alys commented:
Not much progress recently. There's been too much bug support needed.
I'm still working on it when I can.

Alys committed:
Merge remote-tracking branch 'upstream/develop' into admin-panel-2

Alys committed:
add page title to top-level Admin Panel

Also add more details to a router comment (consistent with a similar
comment) in case it helps anyone.

Alys committed:
fix tests

Alys committed:
display Moderation Notes above Contributions

Alys committed:
lint fix

SabreCat commented:
@Alys Is this still going forward? What remains to be done?

Alys commented:
The code so far is working well for us.

As part of the API changes, I added a brief bit of preparation for improving/refactoring how mod/admin permissions are handled in future, thinking that we'd be able to do that slowly in bits and pieces. Matteo wanted it all done in this PR though, and it's a **lot** of work to do all at once so I kind of froze up and couldn't face it - there just aren't enough free hours for me to be able to handle it all as a single project. Sorry. :(  I should probably just remove the initial code that I'd added for that and just leave in the bare minimum of permission changes. I also need to make a small change to the API's admin route to prevent someone without the new permission giving that permission to any account.

SabreCat commented:
That sounds like a good plan! We can tackle a more comprehensive permissions update separately when we want/need to tackle it.

Alys committed:
Merge remote-tracking branch 'upstream/develop' into admin-panel-2

Alys committed:
remove placeholder code for new privileges

I had planned to have each of these implemented in stages, but
paglias wanted it all done at once. I'm afraid that's too big a
project for me to take on in a single PR so I'm cancelling
the plans for adjusting the privileges.

SabreCat commented:
@Alys Ready for another round of testing/review? Or are you still making modifications?

Alys commented:
@SabreCat The PR's current state is suitable for deploying to delta but I still have more code to write. If you wanted to deploy it now, that would reduce support hours and let shanaqui take back more of the support, but I know each additional PR in delta is added work, so if you'd rather wait that's fine. :)

SabreCat commented:
I don't think there's anything else on Delta at the moment, so up it goes!

SabreCat commented:
@Alys can you take a look at these conflicts?

Alys commented:
@SabreCat I think at least some of that conflict will disappear by itself once the Hall's current Display Name duplication issue is fixed in Staging. If there's any conflict left after that, I'll be happy to resolve it, but if I make changes before then, it'll probably result in more conflicts later.

SabreCat commented:
Ah, I thought the duplication was the result of the conflict, not the other way around--
I _think_ I've correctly removed the duplication in 094cf899d8fb51d6c3cee, but the `v-if` behavior was not acting the way I thought it would in the duplicated code to begin with, so I may be confused?

SabreCat commented:
further fixage at 169634f205b2440096b9e514c875fcf0ac3b7246

SabreCat commented:
there was still one more commit after that, but I think now this is the actual conflict needing to be resolved..?

Alys commented:
For my reference, this was the final commit mentioned above: 9e740e70928a5fa313a0dfc2e9cadbc0af849d7f

Alys committed:
revert Hall of Heroes to version from develop

The changes made to this file for the Admin Panel were optional
and only for slight convenience - they allowed an admin to
go to the Hall of Heroes and open any user listed there
into the Admin Panel by clicking a link.
The same result can be achieved by manually copying the UUID or Username
and pasting into the Admin Panel.

Now that the Hall of Heroes has been modified in `develop`,
we can just use that version. There's no real point to
updating it for the optional Admin Panel change,
especially since the bulk of the Hall's admin code will be
removed after the Admin Panel goes live.

SabreCat commented:
Superseded by #13843!
